saleor/checkout/complete_checkout.py

```diff
@@ -180,11 +180,16 @@ def _process_shipping_data_for_order(
     delivery_method_info = checkout_info.delivery_method_info
     shipping_address = delivery_method_info.shipping_address
 
-    if (
+    # allow saving shipping address in case the delivery method is not collection point
+    # or the delivery method is not provided, but the shipping address is set
+    save_shipping_address = checkout_info.checkout.save_shipping_address and (
         delivery_method_info.store_as_customer_address
-        and checkout_info.user
-        and shipping_address
-    ):
+        or not delivery_method_info.delivery_method
+    )
```

Comment:
suggestion: such complex conditions, especially near business logic, would be much more readable if simplified outside of the if statement

such as

```python
address_marked_to_be_saved_via_mutation = checkout_info.checkout.save_shipping_address
delivery_allows_storing_as_address = delivery_method_info.store_as_customer_address
delivery_method_doesnt_exist = delivery_method_info.delivery_method

save_shipping_address = address_marked_to_be_saved_via_mutation 
  and (delivery_allows_storing_as_address or delivery_method_doesnt_exist)

```

which is easier to understand how the logic behaves and easier to find a logical bug

Reply:
I applied suggestion and moved the logic to the separate method